### ISSUE TITLE
Natural Language Processing with Transformers

### ISSUE THREAD

**Issue by sebinsunny** [sebinsunny/blog#7]
# Transformers
To understand transformers, we first need to explain:

* The encoder-decoder framework
* Attention mechanisms
* Transfer learning

## The Encoder-Decoder Framework

Prior to transformers, recurrent architectures were the start of the art in NLP. These architectures contain a feedback loop in the network configuration that allows information to propagate from one step to another, making them ideal for modeling sequential data like text.
![image](https://user-images.githubusercontent.com/8482245/158040695-0b544f66-8238-4e3b-8ee4-d4e543abe982.png)
The job of the encoder is to convert the information from the input sequence into a numerical representation that is often called the hidden state. This state is passed to the decoder, which generates the output sequence.

The weakness of this architecture is the information bottleneck: it has to represent the meaning of the whole input sequence because this is all the decoder has access to when generating the output.
![image](https://user-images.githubusercontent.com/8482245/158041995-908d8ecc-70d9-488f-9755-88e6dcee190f.png)

## Attention mechanisms
Instead of producing a single hidden state for the input sequence, the encoder outputs a hidden state at each step that the decoder can access.
![image](https://user-images.githubusercontent.com/8482245/158042091-e54ef73a-fab8-4a05-8287-b0c2f8bba43e.png)
How to use all the hidden states
![image](https://user-images.githubusercontent.com/8482245/158048102-6473d51d-a81b-47ea-832c-bcb81e24afb9.png)

The decoders previous hidden states, denoted as S_I minus 1, contains information about the previous words in the output translation. This means you can compare the decoder states with each encoder state to determine the most important inputs. Intuitively, the decoder can set the weights such that if it focuses on only the most important inputs words for the next prediction, it decides which parts of the input sequence to pay attention to
![image](https://user-images.githubusercontent.com/8482245/158048307-8899aa19-d42a-4fd4-b6d8-4d440e0eaabe.png)

This is done using the feedforward neural network with the encoder and decoder hidden states as inputs, where the weights for the feedforward network are learned along with the rest of the seek to seek model. The scores are then turned into weights that range from zero to one using the softmax function. This means the weights can be thought of as a probability distribution that sum to one. Finally, each encoder state is multiplied by its respective weights and sum together into one context vector.
